plugins/cpp/parser/src/cppparser.cpp

```diff
@@ -329,19 +332,142 @@ int CppParser::worker(const clang::tooling::CompileCommand& command_)
+                     return std::make_shared<model::File>();
+                   }
+                 });
+
+  try
+  {
+    (util::OdbTransaction(_ctx.db))([&]
+    {
+      for (const model::FilePtr file : filePtrs)
+      {
+        if(file)
+        {
+          markAsModified(file);
+        }
+      }
+
+      for (auto& item : _ctx.fileStatus)
+      {
+        switch (item.second)
+        {
+          /*
+           * In case the file status is MODIFIED or DELETED,
+           * delete the related records to the File entity
+           * from the C++ parser's model.
+           */
+          case IncrementalStatus::MODIFIED:
+          case IncrementalStatus::DELETED:
+          {
+            LOG(info) << "[cppparser] Database cleanup: " << item.first;
+
+            // Fetch file from SourceManager by path
+            model::FilePtr delFile = _ctx.srcMgr.getFile(item.first);
+
+            // Query CppAstNode
+            auto defCppAstNodes = _ctx.db->query<model::CppAstNode>(
+              odb::query<model::CppAstNode>::location.file == delFile->id &&
+              odb::query<model::CppAstNode>::astType == model::CppAstNode::AstType::Definition);
+            for (const model::CppAstNode &astNode : defCppAstNodes)
+            {
+              // Delete CppEntity
+              auto delCppEntities = _ctx.db->query<model::CppEntity>(
+                odb::query<model::CppEntity>::astNodeId == astNode.id);
+              for (const model::CppEntity &entity : delCppEntities)
+              {
+                _ctx.db->erase<model::CppEntity>(entity.id);
+              }
+
+              // Delete CppInheritance
+              auto delCppInheritance = _ctx.db->query<model::CppInheritance>(
+                odb::query<model::CppInheritance>::derived == astNode.mangledNameHash);
+              for (const model::CppInheritance &inheritance : delCppInheritance)
+              {
+                _ctx.db->erase<model::CppInheritance>(inheritance.id);
+              }
+
+              // Delete CppFriendship
+              auto delCppFriendship = _ctx.db->query<model::CppFriendship>(
+                odb::query<model::CppFriendship>::target == astNode.mangledNameHash);
+              for (const model::CppFriendship &friendship : delCppFriendship)
+              {
+                _ctx.db->erase<model::CppFriendship>(friendship.id);
+              }
+
+              // Delete CppNode (connected to CppAstNode)
+              auto delNodes = _ctx.db->query<model::CppNode>(
+                odb::query<model::CppNode>::domainId == std::to_string(astNode.id) &&
+                odb::query<model::CppNode>::domain == model::CppNode::CPPASTNODE);
+              for (model::CppNode &node : delNodes)
+              {
+                _ctx.db->erase<model::CppNode>(node.id);
+                // TODO: handle that RelationCollector will try to reinsert the other end of the relation
+                // Raises "[WARNING] object already persistent"
```

Comment:
Has it been checked that this warning is only a nuisance but does not break actual functionality?

Reply:
Yes. Also, just for the later record, with #274 this will need some refactoring. (But first one of the pull request has to be accepted, so the fix can go into the other one.)